examples/sugarscape/agents.py

```diff
@@ -0,0 +1,219 @@
+import random
+import pylab
+from abce import Agent
+
+
+def get_distance(pos_1, pos_2):
+    x1, y1 = pos_1
+    x2, y2 = pos_2
+    dx = x1 - x2
+    dy = y1 - y2
+    return pylab.sqrt(dx ** 2 + dy ** 2)
+
+
+class Sugar(Agent):
+    def __init__(self, pos, max_sugar):
+        self.pos = pos
+        self.amount = max_sugar
+        self.max_sugar = max_sugar
+
+    def step(self):
+        self.amount = min([self.max_sugar, self.amount + 1])
+
+
+class Spice(Agent):
+    def __init__(self, pos, max_spice):
+        self.pos = pos
+        self.amount = max_spice
+        self.max_spice = max_spice
+
+    def step(self):
+        self.amount = min([self.max_spice, self.amount + 1])
+
+
+class SsAgent(Agent):
+    def init(self, parameters, agent_parameters):
+        self.moore = False
+        self.grid = parameters["grid"]
+        self.set_at_random_unoccupied_pos()
+        self.grid.place_agent(self, self.pos)
+        self.sugar = random.randrange(25, 50)
+        self.spice = random.randrange(25, 50)
+        self.metabolism = random.randrange(1, 5)
+        self.metabolism_spice = random.randrange(1, 5)
+        self.vision = random.randrange(1, 6)
+        self.prices = []
+
+    def set_at_random_unoccupied_pos(self):
+        x = random.randrange(self.grid.width)
+        y = random.randrange(self.grid.height)
+        if not self.is_occupied((x, y)):
+            self.pos = (x, y)
+            return
+        self.set_at_random_unoccupied_pos()
+
+    def get_sugar(self, pos):
+        this_cell = self.grid.get_cell_list_contents([pos])
+        for agent in this_cell:
+            if type(agent) is Sugar:
+                return agent
+
+    def get_spice(self, pos):
+        this_cell = self.grid.get_cell_list_contents([pos])
+        for agent in this_cell:
+            if type(agent) is Spice:
+                return agent
+
+    def get_ssagent(self, pos):
+        this_cell = self.grid.get_cell_list_contents([pos])
+        for agent in this_cell:
+            if isinstance(agent, SsAgent):
+                return agent
+
+    def is_occupied(self, pos):
+        this_cell = self.grid.get_cell_list_contents([pos])
+        return len(this_cell) > 2
+
+    def move(self):
+        # hack this checkalive shouldn't be here
+        if not self.check_alive(): return
```

Comment:
This is a hack because apparently, the event scheduler does not check for dead agents.